packages/protocol/contracts/tokenvault/ERC20Vault.sol

```diff
@@ -227,8 +263,26 @@ contract ERC20Vault is BaseVault {
 
         // If it's a bridged token
         if (bridgedToCanonical[token].addr != address(0)) {
+            // erc20NativeRegistry shall be address(0) on L1 and CAN be address(0) on L2 too, if
+            // there is no native support
+            address erc20NativeRegistry = resolve("erc20_native_registry", true);
+            address adapter;
+            if (erc20NativeRegistry != address(0)) {
+                // Check if it's a native/custom token support
+                (, adapter) =
+                    IERC20NativeRegistry(erc20NativeRegistry).getCanonicalAndAdapter(token);
+            }
+            // Depending on native token (or not), it handles the burn accordingly.
+            _executeBurnOrMint(
+                adapter != address(0) ? adapter : token,
+                token,
+                msg.sender,
+                amount,
+                adapter != address(0),
+                false
+            );
```

Comment:
Very similar code as in ` onMessageRecalled`, I think best if it is shared in a single other function.